src/Oscoin/Data/OscoinTx.hs

```diff
@@ -0,0 +1,331 @@
+    -> WorldState c
+    -- ^ The input world state.
+    -> Either (TxError c) (WorldState c, TxOutput)
+applyTx tx@Tx'{..} author ws = do
+    Account{..} <- lookupAccount author ws
+
+    when (txNonce /= accountNonce) $
+        Left (ErrInvalidNonce txNonce)
+
+    when (txFee < minimumTxFee tx) $
+        Left (ErrInvalidFee txFee)
+
+    when (txFee + txBurn > accountBalance) $
+        Left (ErrInsufficientBalance accountBalance)
+
+    -- TODO(cloudhead): Transfer fee
+    -- TODO(cloudhead): Destroy burn
+    -- TODO(cloudhead): Verify transaction size
+
+    applyTxPayload txPayload author ws
+
+-- | Apply the transaction payload to a world state.
+applyTxPayload
+    :: ( ByteArrayAccess (Crypto.ShortHash c)
+       , CBOR.Serialise (Crypto.ShortHash c)
+       )
+    => TxPayload c
+    -> Address c
+    -> WorldState c
+    -> Either (TxError c) (WorldState c, TxOutput)
+
+applyTxPayload (TxRegisterProject addr) _ ws =
+    if WorldState.member key ws
+       then Left (ErrProjectExists addr)
+       else Right ( WorldState.insert key (ProjectVal (mkProject addr)) ws
+                  , TxOutput )
+  where
+    key = addressKey addr
+
+applyTxPayload (TxUnregisterProject addr) author ws = do
+    p@Project{..} <- lookupProject addr   ws
+    member        <- lookupMember  author ws
+
+    mapHandlerError (pUnregister p member)
+    pure (WorldState.delete projKey ws, TxOutput)
+  where
+    projKey = addressKey addr
+
+applyTxPayload (TxAuthorize addr pk) author ws = do
+    p@Project{..} <- lookupProject addr    ws
+    member        <- lookupMember  author  ws
+    epoch         <- lookupEpoch           ws
+
+    mapHandlerError (pAuthorize pk p member)
+    pure (adjust (addKey epoch) projKey ws, TxOutput)
+  where
+    projKey = addressKey addr
+
+    addKey e v = case v of
+        ProjectVal p -> ProjectVal $
+            p { pMaintainers = Map.insert pk (mkMember pk e) (pMaintainers p) }
+        _ -> v
+
+applyTxPayload (TxDeauthorize addr pk) author ws = do
+    p@Project{..} <- lookupProject addr    ws
+    member        <- lookupMember  author  ws
+
+    mapHandlerError (pDeauthorize pk p member)
+    pure (adjust removeKey projKey ws, TxOutput)
+  where
+    projKey = addressKey addr
+
+    removeKey v = case v of
+        ProjectVal p -> ProjectVal $
+            p { pMaintainers = Map.delete pk (pMaintainers p) }
+        _ -> v
+
+applyTxPayload TxCheckpoint{} _ _ws = notImplemented
+applyTxPayload TxUpdateContract{} _ _ws = notImplemented
+
+applyTxPayload (TxTransfer _ _ bal) _ _ | bal == 0 =
+    Left (ErrInvalidTransfer bal)
+applyTxPayload (TxTransfer sender receiver bal) author ws = do
+    senderAcc   <- lookupAccount sender   ws
+    receiverAcc <- lookupAccount receiver ws
+
+    when (author /= sender) $
+        Left (ErrNotAuthorized author)
+
+    when (bal > accountBalance senderAcc) $
+        Left (ErrInsufficientBalance (accountBalance senderAcc))
```

Comment:
What do you think about throwing these errors as part of the `withdraw` function. Similar throwing `ErrEverflow` as part of `deposit`.

Reply:
Not a bad idea, but not sure it will make the code more readable vs. doing validation upfront, as it will complicate the logic. Happy to explore in Part 2. of this series though.